### DISCUSSION TITLE
How to persist/share the `canvasRef` to child components using a (Recoil) store?

### DISCUSSION THREAD

**Discussion by Vadorequest** [reaviz/reaflow#57] [Q&A]
The documentation at https://reaflow.dev/?path=/story/docs-advanced-refs--page explains a bit how to use the `canvasRef`, but the example is very light and isn't necessarily what a real app would do.

```ts
export const MyCanvas: FC = () => {
  const ref = useRef<CanvasRef | null>(null);

  useEffect(() => {
    // Refs give you ability to do things like:
    // ref.current?.centerCanvas()
  }, [ref]);

  return <Canvas ref={ref} />;
};
```

More specifically, how can I access the `canvasRef` from a component? (Node, Edge, etc.)

I tried to store it in a shared state, but it didn't work out.

```ts
type Props = {
  canvasRef: MutableRefObject<CanvasRef | null>;
}

const CanvasContainer: React.FunctionComponent<Props> = (props): JSX.Element | null => {
  const {
    canvasRef,
  } = props;
  const [persistedCanvasRef, setPersistedCanvasRef] = useRecoilState(canvasRefState);

  useEffect(() => {
    console.log('Persisting canvas ref', canvasRef);
    setPersistedCanvasRef(canvasRef as MutableRefObject<CanvasRef>);
  }, [canvasRef]);

  ...

  return (<Canvas ... />)
}
```

Calling `setPersistedCanvasRef(canvasRef as MutableRefObject<CanvasRef>);` throws the error `TypeError: Cannot assign to read only property 'current' of object '#<Object>'`.

```ts
import { MutableRefObject } from 'react';
import { CanvasRef } from 'reaflow';
import { atom } from 'recoil';

export const canvasRefState = atom<MutableRefObject<CanvasRef> | null>({
  key: 'canvasRefState',
  default: null,
});

```

I thought I could store the canvasRef and allow any component to subscribe to it to access its properties, but it seems I misunderstood something.

Another alternative I see is to pass down the `canvasRef` as props for `Node` and `Edge` components, but that's a bit ugly/heavy.

How would you recommended doing that?

**Comment by gurkerl83** [comment]
@Vadorequest I saw this error, too, putting a ref to a recoil atom. There is a property "dangerouslyAllowMutability" in an atom, maybe also in atomFamily, which allows you to do this kind of thing.

```
export const reparentState = atom({
  key: 'reparent',
  default: initialState,
  dangerouslyAllowMutability: true
});
```

You can find a complete example here. I tried Jotai, which is similar to Recoil; it also has atoms etc., and putting references on those things did work without any error; see the comment in the file.

https://github.com/project-millipede/millipede-docs/blob/14e1444f686cf627b84be1629661e4960abf80ea/packages/demonstrator/navigation/src/recoil/features/reparent/reducers.ts#L25

Maybe this helps, 

Thx!

**Reply by amcdnl** [reply]
I could be wrong but I don't think its ideal to store a Ref/Component in a global store like this.
